### ISSUE TITLE
Exceptions thrown in doOnError are not propagated back when publishOn is used

### ISSUE THREAD

**Issue by dfeist** [reactor/reactor-core#253]
The following scenario causes a hung thread:

```
    Mono.just(1)
        .publishOn(parallel())
        .doOnNext(i -> {
          throw new IllegalStateException();
        })
        .doOnError(e -> {
          throw new RuntimeException(e);
        })
        .block();
```

If the `publishOn()` is removed this behaves as expected and the `block()` exits with exception rather than hanging.

**Comment by dfeist** [comment]
Related to #272

**Comment by dfeist** [comment]
Just confirmed this passes with fix for #272.  Tested with this:

```
    AssertSubscriber<Integer> assertSubscriber = new AssertSubscriber<>();

    Mono.just(1)
        .publishOn(parallel())
        .doOnNext(i -> {
          throw new IllegalArgumentException();
        })
        .doOnError(e -> {
          throw new IllegalStateException(e);
        })
        .subscribe(assertSubscriber);

	assertSubscriber
            .await()
            .assertError(IllegalStateException.class)
            .assertNotComplete();
```
